File: rogue-js/src/js/rogue/element/monster/comportement/random-walk.js
import { createGoTo } from "js/rogue/element/monster/comportement";
import { getCircle } from "js/rogue/tools";

class RandomWalk {
  constructor(monster) {
    this.monster = monster;
    this.nx = monster.x;
    this.ny = monster.y;
    this.goTo = createGoTo(this.monster, monster.x, monster.y);
    this.reset();
  }

  checkNextDir(world) {
    let dx = this.monster.x;
    let dy = this.monster.y;

    let ray = this.monster.depht;
    let find = false;
    while (!find) {
      const points = getCircle(this.monster.x, this.monster.y, ray);
      while (points.length > 0) {
        const p = points.pop();

        if (world.elementCanGoTo(this.monster, this.monster.x, this.monster.y, p.x, p.y)) {
          find = true;
          dx = p.x;
          dy = p.y;
          break;
        }
      }
      if (!find) {
        ray--;
      }
    }
    this.goTo.reset(dx, dy);
  }

  reset() {
    this.nx = this.monster.x;
    this.ny = this.monster.y;
  }

  activate(world) {
    if (this.monster.speed > 0) {
      if (this.goTo.isFinished()) {
        this.checkNextDir(world);
      } else {
        this.goTo.activate(world);
      }
    }
  }

  isFinished() {
    return false;
  }
}

export default monster => new RandomWalk(monster);

File: my-rogue - 2/src/main/java/com/renaud/rogue/element/comportement/RandomWalk.java
package com.renaud.rogue.element.comportement;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;

import com.renaud.rogue.element.Monster;
import com.renaud.rogue.game.Game;
import com.renaud.rogue.tools.MathTools;
import com.renaud.rogue.tools.Point;
import com.renaud.rogue.world.World;

public class RandomWalk implements Comportement {

    int nx;
    int ny;

    Monster monster;

    GoTo goTo;

    public RandomWalk(Monster monster) {
	this.monster = monster;
	goTo = new GoTo(monster, monster.getX(), monster.getY());
	reset();
    }

    private void checkNextDir(World world) {
	int dx = monster.getX();
	int dy = monster.getY();
	int ray = monster.getDepht();
	boolean find = false;
	while (!find) {
	    List<Point> points = new ArrayList<>(MathTools.getCercle(monster.getX(), monster.getY(), ray));
	    Collections.shuffle(points, new Random());
	    while (!points.isEmpty()) {
		Point p = points.remove(0);
		if (world.canGo(monster.getX(), monster.getY(), p.getX(), p.getY())) {
		    find = true;
		    dx = p.getX();
		    dy = p.getY();
		    break;
		}
	    }
	    if (!find) {
		ray--;
	    }
	}
	// System.out.println(monster.getX() + " " + monster.getY() + " " + dx + " " +
	// dy);
	goTo.reset(dx, dy);
    }

    @Override
    public void reset() {
	nx = monster.getX();
	ny = monster.getY();
    }

    @Override
    public void activate(Game game) {
	// if (!monster.turnIsEnd()) {
	if (goTo.isFinished()) {
	    checkNextDir(game.getWorld());
	    this.activate(game);
	} else {
	    goTo.activate(game);
	}
	// }
    }

    @Override
    public boolean isFinished() {
	return false;
    }
}
